### ISSUE TITLE
Attachment queueing?

### ISSUE THREAD

**Issue by hotmeteor** [CodeSleeve/stapler#96]
Is queueing built into the attachment processing? A case study in mind:
1. Select an image to be attached to an Item
2. The image is large, and requires a few different sizes to be generated
3. A config determines the size to generate first (perhaps a thumbnail) and then the rest of the processing is passed off to a queue so the user doesn't have to wait

I'm not sure if anything like this exists in the current package. Thanks for any insight.

**Comment by hotmeteor** [comment]
I've not done multiple files, but the way I've handled this is basically the following:

1. Upload the original file(s) with Stapler
- Optionally create one thumbnail to indicate the image was uploaded (on the next screen or something)
- Queue up some processes to go "reprocess" those files using a different config than the original upload config

**Comment by jorenvanhee** [comment]
I made a rough version of Stapler with queueing that I'm using in a project. Most of the queueing stuff is done in laravel-stapler. I would like to share and explain my code if you're interested.

**Comment by tomvo** [comment]
I would be interested, still something on my todo list for one of our products that uses stapler intensively.

**Comment by jorenvanhee** [comment]
Ok, this is what I'm working on right now.

### stapler
#### defer & defer_all_except options
```php
$this->hasAttachedFile('picture', [
  'styles' => [/* your styles */],
  'defer' => [/* styles you don't want to process immediately */],
  'defer_all_except' => [/* opposite of 'defer' option */]
]);
```

#### reprocess specific styles
```php
$attachment->reprocess($styles = []);
```

#### process deferred styles
Still have to add a method to process only the deferred styles. Will use the reprocess method.

### laravel-stapler
#### ProcessDeferredAttachmentStyles Job
This job processes the deferred attachment styles.

#### QueueingEloquentTrait
QueueingEloquentTrait will automatically dispatch the ProcessDeferredAttachmentStyles when needed.

---

Let me know what you think.

**Comment by tomvo** [comment]
@jorenvanhee have you managed to get your queuing solution working?

**Comment by jorenvanhee** [comment]
@tomvo Yep, got it working. I'm using my own forks now. But I would be happy to clean it up, document it, and submit a PR (if there's interest).

Here's my fork of [stapler](https://github.com/jorenvanhee/stapler/tree/queue) and [laravel-stapler](https://github.com/jorenvanhee/laravel-stapler/tree/queue).

The word queue was already used in the codebase, so I used "defer". You'll see it in my commits.

**Comment by tomvo** [comment]
@jorenvanhee I checked your code and it actually looks pretty solid. The changes are actually not even that big. However it's Laravel 5 only. Unfortunately my legacy codebase is based on L4 so I wouldn't be able to help you out right away with testing it. I could adapt it to L4 though.

**Comment by tabennett** [comment]
I'm working on a 2.0 release of Stapler that's php7 and laravel 5.1 only. It's got some new features:
* Flysystem for file abstraction (Local, S3, Rackspace, and Azure Cloud File storage)
* Easy image naming/renaming
* Dynamic style creation: ```$image->picture->url('medium', '150x150')```
* Possibly more drivers/traits (Cakephp, Doctrine, Analogue, Yii, etc), however this is stretch goal.
* More examples (example applications, etc) and better documentation
* Better/more test coverage
* Several bug fixes, etc.

I'm going to release one more minor release from the 1x branch and back port as much of this stuff as possible to, but after that It's going to be bug fixes only for that branch. I'd really like to figure out a clean way to do attachment queueing that (works nicely/elegantly with this package) and get that into the new release for sure. I took a look at that code and it looks pretty clean, but there will more than likely have to be changes/tweaks to make it work consistently with the rest of the package. If you guys want to work on something bounce ideas around I think we could get it done.

**Comment by tomvo** [comment]
Great that you're working on v2 of the library. I guess it wouldn't make much sense for the package to add queuing support for L4. @jorenvanhee if you don't mind i'm going to use your `stapler` fork and use your ideas on the laravel fork and port it to a L4 version. I'll share it back here when done for anybody else who might benefit from this.

**Comment by jorenvanhee** [comment]
@tomvo Great, go ahead!
